### ISSUE TITLE
Resize filter not returning new path

### ISSUE THREAD

**Issue by saralohr** [timber/timber#1303]
### Expected behavior
**Resize filter should return new path**
I'm experiencing a similar issue to [#237](https://github.com/timber/timber/issues/237). Resize filter was working, but now it is not. The image is still being created (it is in the `content/uploads` directory), but the original upload URL is being included instead. Oddly, it is working just fine on my staging site and local server with the same codebase and Timber / WP versions. I've had this happen recently with another site too, but the error occurred on the staging site rather than production.

I noticed the issue after upgrading to Timber 1.2.1, but I am using the same version of Timber across all of my installs and not experiencing the problem everywhere.

Questions from Issue #237:

> Are you running WP from a subdirectory? 

No
> Do you have "normal" setup with /wp-content/uploads/ etc.? 

Yes
> Do all images fail? 

Yes

### Actual behavior
Resize filter creating cropped image but not returning the path

### Steps to reproduce behavior
```
{% for accessory in post.get_field('featured_accessory') %}
  <img alt=""
    src="{{ TimberImage(accessory.image).src | resize(400, 266) }}"
    sizes="(max-width: 47.9375em) 100vw, (max-width: 59.9375em) 50vw, 33vw"
    srcset="{{ TimberImage(accessory.image).src | resize(400,266) }} 400w, {{ TimberImage(accessory.image).src | resize(750,500) }} 750w">
{% endfor %}
```

### What version of WordPress, PHP and Timber are you using?
WordPress 4.7.1,  Timber 1.2.1

### How did you install Timber?
Upgraded to newest version via plugin updater in WordPress dashboard.

**Comment by saralohr** [comment]
Um... so this problem seems to have fixed itself after WPEngine cache purge. Closing!

**Comment by timknight** [comment]
I can confirm this issue too. Also on WPEngine the issue however seems to be strictly over SSL for when it's generating a path that includes https.

**Comment by timknight** [comment]
Digging in further I'm seeing this isn't just a WPEngine issue since both @saralohr and I seem to be WPEngine users. 

With debug mode enabled I get a hard error:

> **Notice**: Undefined index: dirname in **[...]/wp-content/plugins/timber-library/lib/ImageHelper.php** on line **383**
> 
> **Notice**:  Undefined index: extension in **[...]/wp-content/plugins/timber-library/lib/ImageHelper.php** on line **385**

With the debug mode disabled I just get the default path over HTTPS.

With HTTPS: `https://timber.dev/wp-content/uploads/2017/01/z55cr_d0ayg-andrew-neel.jpg`
With HTTP: `http://timber.dev/wp-content/uploads/2017/01/z55cr_d0ayg-andrew-neel-600x0-c-default.jpg`

**Comment by saralohr** [comment]
I also was wondering if this was an SSL issue, as I am using `https` on production but not on staging. However, the other site where I experienced the same issue a couple of months ago does not use SSL, and I am experiencing the issue on staging but _not_ production. I just double checked the Timber version on that site and it is still `1.1.12`.

**Comment by timknight** [comment]
So you're experiencing the issue on `1.1.12` or that's the production system where it's not happening?

**Comment by saralohr** [comment]
Sorry, so many to keep track of. It is happening on two sites:

### Site number 1
+ Issue on PRODUCTION: 
  + `https`
  + Timber version `1.2.1`
  + reverting to Timber `1.1.3` fixes the issue
+ **Not** experiencing the issue on STAGING
  + `http`
  + Timber version `1.2.1`
 
### Site number 2
+ Issue on STAGING: 
  + `http`
  + Timber version `1.1.12`
+ **Not** experiencing the issue on PRODUCTION
  + `http`
  + Timber version `1.1.12`

**Comment by timknight** [comment]
Okay interesting. My situation is like your **site number 1**. Except reverting to 1.1.12 fixes the issue just fine too.

**Comment by saralohr** [comment]
BTW, both of these use WPEngine, no CDN.

To make this more interesting, yet another site I work on uses WPEngine _with_ a CDN, recently upgraded to Timber `1.2.1`, as is **NOT** experiencing the issue.

**Comment by timknight** [comment]
No CDN on mine either. 

If you enable WP_DEBUG do you get the same or a different error message as the one I posted? You have to view source since it's within the img tag so it doesn't actually render.

**Comment by saralohr** [comment]
I haven't been able to replicate the issue locally, unfortunately, so I'm
not getting any error messages.

**Comment by timknight** [comment]
What about one of your staging sites? Can you put those in debug mode? Also are you using free ACF or ACF Pro?

**Comment by jarednova** [comment]
@timknight @saralohr — thanks for the reports on these. I found some stuff in #1293 that deals with the errors @timknight is seeing with...

> Notice: Undefined index: dirname in [...]/wp-content/plugins/timber-library/lib/ImageHelper.php on line 383
> Notice: Undefined index: extension in [...]/wp-content/plugins/timber-library/lib/ImageHelper.php on line 385

... I'm hoping solving those might relate here

**Comment by timknight** [comment]
Thanks @jarednova 👍

**Comment by saralohr** [comment]
Thanks @jarednova! 

@timknight I turned on `DEBUG` on the staging site where I am seeing the issue and didn't get an error. I noticed something else funny happening, though. The images where it is happening (it's not all of them), redirect to the production URL when I open them in a new tab. They have the staging path in the page though. 

> Also are you using free ACF or ACF Pro?

I am using ACF Pro.

**Comment by daviddarke** [comment]
+1 getting @saralohr issue from her example 'Site number 1'

**Comment by Flowgram** [comment]
I've got the exact same issue on WPEngine.

Timber version 1.2.3.

It doesn't work on https Production or Staging. It works on http Staging though.

**Comment by simonlewisjones** [comment]
Can confirm same issue on WPEngine:
Staging http works.
Production http works.
Production https doesn't.

Timber version 1.2.4.

**Comment by timknight** [comment]
Still seems this is an issue for me as well. I'm still seeing this over https in general (production or staging) anytime the resize filter is used. 

So with:

```
<img src="{{ post.thumbnail.src('large') }}" alt="{{ post.thumbnail.title }}" class="img-responsive">
```

I get:

```
<img src="https://dev.site.com/wp-content/uploads/2017/03/david-cohen-180217-1024x682.jpg" alt="david-cohen-180217" class="img-responsive">
```

But with:

```
<img src="{{ post.thumbnail.src|resize(1000) }}" alt="{{ post.thumbnail.title }}" class="img-responsive">
```

I get:

```
<img src="https://dev.site.com/wp-content/uploads/2017/03/david-cohen-180217.jpg" alt="david-cohen-180217" class="img-responsive">
```

If I use the resize filter over http, I get:

```
<img src="http://dev.site.com/wp-content/uploads/2017/03/david-cohen-180217-1000x0-c-default.jpg" alt="david-cohen-180217" class="img-responsive">
```

**Comment by timknight** [comment]
So I spent some time digging into this today, you know, just for "fun".

Okay, so stupid question. Those of you having the issue, in your General Settings do you have your site address and your WordPress with the `https:` in the address?

tl;dr; - make sure your site address and wordpress address have the https in it.

Here's what it looks like is happening to me @jarednova.

The `analyze_url` function within `ImageHelper.php` uses the `wp_upload_dir()` to determine if the image is within the uploads folder. If you're going to "https" in the browser while the settings still have "http", the check `strpos($tmp, $upload_dir['baseurl'])` will fail since it can't find the "https" string within `$upload['baseurl']` that will then take it down to the other condition used for content-based items where it fails as well. 

This then sends it back up to `_operate` where it checks `$op->run($source_path, $destination_path)` and fails (because the final condition within `analyze_url` for wp-content doesn't construct a real subdir from `theme_url_to_dir` when it's actually an upload file). So after all that _operate just returns the source hitting the `// in case of error, we return source file itself` part of the condition.

**Comment by joeldbirch** [comment]
Thanks @timknight, I was experiencing this issue (https, WP Engine, with CDN, Timber 1.2.4) and (after reading your comment) noticed the WordPress path was set to "http", while the site address was "https". Updating the WordPress path to "https" fixed the issue.

**Comment by Flowgram** [comment]
My entire multisite had http as site URLs instead of https. Problem now solved for me.

Thanks guys.

**Comment by tremby** [comment]
I have the same issue. I'm on HTTPS, and those two settings are greyed out for me, but do start `https:`. I'm using Trellis and Bedrock.

I'm on Timber 1.3.0. If I roll back to 1.2.4, the problem disappears.

**Comment by jarednova** [comment]
Thanks so much everyone (especially @timknight who left all the breadcrumbs here). I just submitted a PR to hopefully tackle this one. Tests are running right now — but try applying it to your install and see if it resolves!

**Comment by tremby** [comment]
This does not fix the issue for me. I will now bisect and try to find the regression which is affecting me.

**Comment by jarednova** [comment]
Ugh, sorry to hear that @tremby — but bisecting would be incredibly helpful so we can pinpoint what caused this. Thanks for looking further into it!

**Comment by tremby** [comment]
```
73bd23ef8b2005b5e9db7122e61415cf55bf9798 is the first bad commit
commit 73bd23ef8b2005b5e9db7122e61415cf55bf9798
Author: gillesgoetsch <gilles.goetsch@me.com>
Date:   Fri Mar 3 19:44:45 2017 +0100

    [Fix] Wrong image path breaking timber resize when using external wp-content-dir (WP_CONTENT_DIR)

:040000 040000 7e084ac9830f221b49aeee722a6b7db3e298d108 8508891e48f45096783a2d6de7085ee0e5484e14 M	lib
```

Investigating further...

**Comment by tremby** [comment]
OK, in [this chunk of code](https://github.com/timber/timber/blob/1303/url_helper_starts_with/lib/ImageHelper.php#L370-L383)

```php
			if ( false === strpos($upload_dir['baseurl'], site_url()) ) {
				// use HOME_URL and relative image path
				$tmp = get_home_url().$tmp;
			} else if ( !$result['absolute'] ) {
				$tmp = site_url().$tmp;
			}
			if ( URLHelper::starts_with($tmp, $upload_dir['baseurl']) ) {
				$result['base'] = self::BASE_UPLOADS; // upload based
				$tmp = str_replace($upload_dir['baseurl'], '', $tmp);
			} else if ( URLHelper::starts_with($tmp, content_url()) ) {
				$result['base'] = self::BASE_CONTENT; // content-based
				$tmp = self::theme_url_to_dir($tmp);
				$tmp = str_replace(WP_CONTENT_DIR, '', $tmp);
			}
```

I have

- `$upload_dir['baseurl']`: `https://ceso.test/app/uploads`
- `site_url()`: `https://ceso.test/wp`
- `$tmp`: `https://ceso.test/app/uploads/2016/03/1.png`

and so we enter the first branch of that conditional. My `get_home_url()` is `https://ceso.test` so this sets `$tmp` to `https://ceso.testhttps://ceso.test/app/uploads/2016/03/1.png` which is obviously no good.

Neither branch of the second conditional is entered: the `$tmp` string does not start with my upload dir baseurl, nor does it start with my `content_url()`, which is `https://ceso.test/app`.

My directory layout, as you can see, is non-standard. It's what is set out by [Roots Bedrock](https://github.com/roots/bedrock), which is a pretty popular thing.

**Comment by krisscox** [comment]
Hello, another Bedrock user here experiencing the same problem as outlined above, $tmp becomes:

http://foo.devhttp://foo.dev/app/uploads/2017/05/bar.jpg

As a hacky fix as I don't have the time to look through the code and comprehend it, I added the following:

```PHP
if ( false === strpos($upload_dir['baseurl'], site_url()) ) {
    // use HOME_URL and relative image path
    if (false === strpos($tmp, get_home_url())) {
        $tmp = get_home_url().$tmp;
    }
} else if ( !$result['absolute'] ) {
    $tmp = site_url().$tmp;
}
```

I'd make a pull request but it would likely do more harm than good!

**Comment by DarrenTheDev** [comment]
Hi all,

I'm experiencing the same issue on multiple sites with version 1.3.0. Reverting to 1.2.4 eliminates the issue.

We're using non-standard site and content urls, configured with constants. So the cause seems to be the same as the one identified by @tremby.

**Comment by lmartins** [comment]
Also having the same issue with similar circumstances. HTTPS site hosted on WPEngine, full sized images served to non logged in users and under HTTPS. Logged in users see resized images, without https everyone gets the resized images, so the problem seems to result from the combination of https and logged off.

**Comment by jarednova** [comment]
Thanks everyone for your reports. There are a few different circumstances listed here, but the most frequent info seems to be regarding Timber 1.3.0 and Roots/Bedrock (@krisscox, @tremby). I was able to get a Bedrock/Timber instance up-and-running and confirmed that v 1.3.0 broke resizing, but luckily also confirmed that v1.3.1 fixed it. 

The issue that @lmartins is referring to seems distinct, so I'm going to open a new thread to focus on that.
